cpp/src/slice2java/JavaUtil.cpp

```diff
@@ -2338,6 +1905,128 @@ Slice::JavaGenerator::createOutput()
+    };
+    knownMetadata.emplace("java:buffer", std::move(bufferInfo));
+
+    // "java:getset"
+    MetadataInfo getsetInfo;
+    getsetInfo.validOn = {typeid(ClassDecl), typeid(Slice::Exception), typeid(Struct), typeid(DataMember)};
+    getsetInfo.acceptedArgumentKind = MetadataArgumentKind::NoArguments;
+    knownMetadata.emplace("java:getset", std::move(getsetInfo));
+
+    // "java:implements"
+    MetadataInfo implementsInfo;
+    implementsInfo.validOn = {typeid(ClassDecl), typeid(Struct)};
+    implementsInfo.acceptedArgumentKind = MetadataArgumentKind::RequiredTextArgument;
+    implementsInfo.mustBeUnique = false;
+    knownMetadata.emplace("java:implements", std::move(implementsInfo));
+
+    // "java:package"
+    MetadataInfo packageInfo;
+    packageInfo.validOn = {typeid(Unit), typeid(Module)};
+    packageInfo.acceptedArgumentKind = MetadataArgumentKind::SingleArgument;
+    packageInfo.extraValidation = [](const MetadataPtr&, const SyntaxTreeBasePtr& p) -> optional<string>
+    {
+        // If 'java:package' is applied to a module, it must be a top-level module.
+        if (auto mod = dynamic_pointer_cast<Module>(p))
+        {
+            // Top-level modules are contained by the 'Unit'. Non-top-level modules are contained in 'Module's.
+            if (!dynamic_pointer_cast<Unit>(mod->container()))
+            {
+                return "the 'java:package' metadata can only be applied at the file level or to top-level modules";
+            }
+        }
+        return nullopt;
+    };
+    knownMetadata.emplace("java:package", std::move(packageInfo));
+
+    // "java:serializable"
+    MetadataInfo serializableInfo;
+    serializableInfo.validOn = {typeid(Sequence)};
+    serializableInfo.acceptedArgumentKind = MetadataArgumentKind::RequiredTextArgument;
+    serializableInfo.extraValidation = [](const MetadataPtr&, const SyntaxTreeBasePtr& p) -> optional<string>
+    {
+        if (auto seq = dynamic_pointer_cast<Sequence>(p))
+        {
+            auto builtin = dynamic_pointer_cast<Builtin>(seq->type());
+            if (!builtin || builtin->kind() != Builtin::KindByte)
+            {
+                return "the 'java:serializable' metadata can only be applied to byte sequences (`sequence<byte>`)";
+            }
+        }
+        // This metadata conflicts with 'java:type' and 'java:buffer', but we let the validation functions for those
+        // metadata emit the conflict warnings instead of this one, since this can only go on the definition.
+        return nullopt;
+    };
+    knownMetadata.emplace("java:serializable", std::move(serializableInfo));
+
+    // "java:serialVersionUID"
+    MetadataInfo serialVersionUIDInfo;
+    serialVersionUIDInfo.validOn = {typeid(ClassDecl), typeid(Slice::Exception), typeid(Struct)};
+    serialVersionUIDInfo.acceptedArgumentKind = MetadataArgumentKind::SingleArgument;
+    knownMetadata.emplace("java:serialVersionUID", std::move(serialVersionUIDInfo));
```

Comment:
should we do the validation `serialVersionUIDInfo` here in an extraValidation function?

Reply:
Good idea!
I moved the extra `stoll` check into the metadata validation, instead of doing it on the spot.